### PR TITLE
Renames class to Sequence

### PR DESCRIPTION
**Pull request by DidierMalenfant**
Prevents name clashes with variables like in NobleEngine.

This is up to you whether you prefer to keep it the old way or not...

### CODE REVIEW

**Comment by DidierMalenfant** [comment]
Hoping the v1.0.1 tag will make it thru too 😂

**Comment by DidierMalenfant** [comment]
I've seen uses of the class as `sequence` in other places. Best to keep it the way you intended. I'll make the changes in NobleEngine instead.

**Comment by Mark-LaCroix** [comment]
Just seeing this now. Noble Engine uses an old version of this library that predates its appearance on GitHub. I renamed it to "Sequence" to conform to the engine's strict naming/style conventions. @DidierMalenfant, I don't think I got a PR from you, but I see the argument to reverting my change in the engine. I'm sorta torn on how to proceed:

- I could bring in the current version of the library and change it to "Sequence" as before.
- I could use an alias (`Sequence = sequence`).
- I could leave it as "sequence" and just let it be "wrong" from the perspective of the engine. It's a third-party library.

The thing preventing me from doing the last two is that I am also concerned about user-created name collisions, since `sequence = sequence.new()` would be an easy mistake for a new developer to make (and hard to debug). I suppose the best practice in most cases (especially inside a Noble Engine project) would be `local sequence = sequence.new()` or `self.sequence = sequence.new()` which I think wouldn't create that problem(?).

Thoughts? I wouldn't expect @NicMagnier to entertain reopening this PR _just_ for the benefit of Noble Engine's OOP-y style, but I also would like to hear his opinion/recommendations since I'm making use of his work.

**Comment by DidierMalenfant** [comment]
Sorry about the lack of PR on my end. Interest in toybox.py, both from Panic and from the community, has been pretty low and so I drifted slowly toward my fantasy Analogue Pocket console project instead.

Personally I still think this class should be named 'Sequence' to visually separate class names from, for exemple, variables.

Regardless of the naming thought, I remember that the issue I was encountering was related to libraries like this inside Noble. Ideally, the "Right Way™️" of including those would be to add them to Noble via a toybox too, i.e. not include them inlined into the Noble Repo.

Problem with that is that people would need toybox.py to use Noble. If the adoption of toybox.py was wide that may be the right way to go but given the current situation I'm not sure.

**Comment by Mark-LaCroix** [comment]
> Sorry about the lack of PR on my end

No worries whatsoever. I know a thing or two about letting things sit around for awhile. 😆

> Problem with that is that people would need toybox.py to use Noble.

Yeah! It's a tricky conundrum. Toybox is so cool and a great way to bring a bunch of stuff together, but Noble Engine (read: me) has a really strong opinion about "The Right Way," which is to reject multi-step development environment setups and cut down on multi-layer dependencies entirely if possible. I've used too many FOSS tools where I spent 100,000x more time hunting down missing documentation, watching YouTube tutorials for a dependency three-levels deep, and Googling terminal script errors than executing any commands, only to discover when I got it working that all it did was put some files in some folders.

My first thought for Sequence was to nest it (and any other 3rd party library) as a git submodule, because submodules are a first-class git feature, simple to set up, and play nice with GUI git clients, but I'm getting pushback from users who are confused by my use of it in the engine's sample project template, and honestly I can't blame them (what I wouldn't give for `--recurse-submodules` to be the default flag when running `git clone`!).

I do still want to get Noble Engine on Toybox (because I don't want to exclude people who want to set up their projects that way), but I can't force Noble Engine users into _any_ system that requires them to configure much more than the Playdate SDK to get started. I'm plenty comfortable in a terminal these days, but I started as a graphic artist and Flash dev, so I know what it's like to feel left out of the "real programmer" club because I didn't have what are basically IT skills, so I always want this stuff to be available but optional.

You're right, though, if Toybox was a more expected part of the standard Playdate developer experience, I'd be way more comfortable hitching Noble Engine to it. It's not really fair to Toybox, honestly. Chicken and egg stuff.

> I drifted slowly toward my fantasy Analogue Pocket console project instead.

Oooo, sounds way more fun than all the nonsense I'm working on!

**Comment by DidierMalenfant** [comment]
I've heard horror stories about submodules... That's why toybox doesn't use them when it imports dependencies. It brings all the files into the local project so that, as you mention, everything is in the user's repo at all times no matter what happens to the dependencies.

There may be a half way solution, which would be to have Noble use toybox to bring Sequence in then check in the result in the Noble repo to that toybox is not needed to use Noble.

I haven't tried that but I get the feeling we may hit a problem with the way `pdc` works in order to make this work. That's also one of the reason I gave up, some things are just not flexible on the SDK side, parts of it broke on an almost regular basis and I didn't get the feeling that this was important to Panic, which is 100% their call of course.

**Comment by Mark-LaCroix** [comment]
> It brings all the files into the local project so that, as you mention, everything is in the user's repo at all times no matter what happens to the dependencies.

That's not how submodules work. They _do_ import the files into a user's working directory, but they aren't _tracked_ by the "parent" repo. Instead, the submodule's folder in the project's working folder also holds it's own .git file which has a reference to the specific commit that the submodule is synced to. The files themselves act as if they're listed in your .gitignore file (but you don't need to configure your .gitignore).

The idea is that when the submodule's repo is updated by its author, you can pull it on your end independently from your own code if you want to, but you also don't have to if you don't want to, since your project knows which commit of the submodule to use. When you do pull a new version, it just becomes a one-line commit to the submodule's .git file in your repo, so the history of which version of the submodule you are using is always tracked, too, which really helps if you need to turn back the clock on your project when either your code or a submodule introduced a breaking change.

This is what makes submodules great for development, too. A lot of projects (like this one) end up having extra files around them in order for the author to properly test them in development, but that's kinda tedious and a waste of resources. For Noble Engine, I have a test project repo that Noble Engine is just a submodule of, so I can muck around with this project when testing/using the engine, but when I make changes to the engine's code, I just push commits on the submodule, so I don't need to include the superfluous main.lua, pdxinfo, and/or whatever else that I need in order to test my work.

I often work on the engine inside multiple games at the same time (particularly when I'm reviewing other people's PRs). The various instances of the engine never get out of sync with each other because I push/pull the submodule (and the branch I'm working with) when I switch to a different project.

Likewise, when a user clones the Noble Engine project template, they aren't stuck on the version of the engine that they started with (which is important as it's still pre-1.0.0).

Basically, it's a package manager built directly into git. If you use git on the command line, you can use it that way (so it's _really_ like a package manager), but if you're using a git GUI client like Fork, GitHub Desktop, SourceTree, you can just double-click the submodule and it opens up in a new tab as its own repo, where you can inspect and pull available changes (as well as submit PRs, or push if you have access).

Anyway, I just find them incredibly easy to use and (vitally) it requires no extra software or tools to install or learn. I do think they're not widely used, though, which creates the same problem any other solution has, which is discoverability/familiarity.

Anyway, thank you for coming to my TED talk. 🤣

**Comment by DidierMalenfant** [comment]
> That's not how submodules work. 

Yeah... I know 😄.. that was in reference to how toybox.py works instead.

I too was using submodules for a while but ended up giving up on them. There are some merging/scalling issues of course but to me at the end of the day the main reason was to get back to truly guaranteed atomic commits on my projects.

If you use a submodule and the other repo goes dark, other people can't use your project anymore. If the other repo re-bases and wipes out commits, your own commit tree is hosed.

So toybox.py was an exercise in trying to get the best of both worlds. If you bring in a toybox to your project it is local, it integral to your repo and can never be messed up. It's synced to the same commits are your project. At the same time, you can update the toybox and that will grab the latest version of it from the other repo.

That's the idea anyway...
